FIR IDE: Implement HLAddNameToArgumentIntention

punzki opened the pull request:
Supporting HL API changes are at https://github.com/JetBrains/kotlin/pull/4502.

punzki committed:
Add all Gradle build output to .gitignore

punzki committed:
FIR IDE: Generate HL intention tests for addNameToArgument.

punzki committed:
FIR IDE: Implement HLAddNameToArgumentIntention.

punzki committed:
FIR IDE: Rename tests for MixedNamedArgumentsInTheirOwnPosition.

TODO: Figure out why FE 1.0 tests are ignoring LANGUAGE_VERSION and
language settings directive.

punzki committed:
FIR IDE: Support annotation calls and delegated constructor calls in
HLAddNameToArgumentIntention.

punzki committed:
FIR IDE: Add description for HLAddNameToArgumentIntention (copied from
FE 1.0 implementation).

punzki commented on idea/tests/testData/intentions/addNameToArgument/mixingPositionedAndNamedDisabled.kt:
```diff
@@ -0,0 +1,8 @@
+// IS_APPLICABLE: false
+// COMPILER_ARGUMENTS: -XXLanguage:-MixedNamedArgumentsInTheirOwnPosition
```
@darthorimar I could not figure out why the FE 1.0 intention tests are ignoring these language settings. Right now it fails. Even `// LANGUAGE_VERSION: 1.3` gets ignored; in a debugger I can see the version is 1.5.

darthorimar replied:
Probably, this is because `AbstractIntentionTest` does not support such directives. You may override`getProjectDescriptor` in `AbstractIntentionTest` and add needed behavior there

punzki replied:
That does not seem to be the cause. The directives are being read in the call to [`withCustomCompilerOptions()`](https://github.com/JetBrains/intellij-kotlin/blob/09b296ff5b52d77e5b66c87b3a509c6303cad9b4/idea/tests/test/org/jetbrains/kotlin/idea/intentions/AbstractIntentionTest.kt#L101). The problem is the preceding call to `myFixture.configureByFiles()` actually causes resolution on FE 1.0 but not on FIR. The resolution causes the language version settings for the project and module to be cached, and this happens before the custom compiler options were read. Moving that `configureByFiles()` to inside `withCustomCompilerOptions()` fixes the issue (see latest commit).

As to why resolution happens for FE 1.0, I dug into it a little bit and the resolution happens here [in `KotlinLanguageInjectionContributor`](https://github.com/JetBrains/intellij-kotlin/blob/09b296ff5b52d77e5b66c87b3a509c6303cad9b4/jvm/src/org/jetbrains/kotlin/idea/injection/KotlinLanguageInjectionContributor.kt#L277) at some point during the fixture setup. Specifically, I think it happens because the caret is inside a string literal and some kind of check happens for injection contributors. `KotlinLanguageInjectionContributor` is only present in the FE 1.0 plugin.

jsjeon commented on fir/src/org/jetbrains/kotlin/idea/fir/intentions/HLAddNameToArgumentIntention.kt:
```diff
@@ -0,0 +1,102 @@
+                return@inputProvider null
+            }
+
+            val varargArgumentCount = resolvedCall.argumentMapping.values.count { it == valueParameterSymbol }
+            if (varargArgumentCount != 1) {
```
Are we going to check if this `vararg` appears only and exactly once?

punzki replied:
We can only add the parameter name for an argument for a vararg parameter if it's the ONLY argument for the parameter. E.g.,:
```kt
fun foo(vararg i: Int) {}

foo(1, 2) // Can NOT add `i = ` to either argument
foo(1) // Can change to `i = 1`
```
Adding a comment to the code.

jsjeon replied:
Maybe add those examples to code comments as well.

punzki replied:
Of course, I should have! Done.

jsjeon commented on fir/src/org/jetbrains/kotlin/idea/fir/intentions/HLAddNameToArgumentIntention.kt:
```diff
@@ -0,0 +1,102 @@
+                val name = input.name
+
+                val prevSibling = element.getPrevSiblingIgnoringWhitespace()
+                if (prevSibling is PsiComment && """/\*\s*$name\s*=\s*\*/""".toRegex().matches(prevSibling.text)) {
```
Just to double-check. Don't we have any convention regarding where/how to put regex?

punzki replied:
It's not a constant regex. It includes `$name`. Otherwise I'd put it in a companion object.

jsjeon commented on idea/tests/testData/intentions/addNameToArgument/annotationCall.kt:
```diff
@@ -0,0 +1,4 @@
+annotation class A(val p1: Int, val p2: Int)
+
+@A(1, <caret>2)
```
Maybe a dumb question, but is there any special reason that we suggest to use named argument for `p2` over `p1`?

punzki replied:
I didn't intend anything with it; I copied `superClassConstructor.kt` and just changed it slightly. 😛 . But I suppose this makes sure it's picking the correct mapped argument and not just the first one, or the first one that matches the type.

jsjeon replied:
It's still unclear, since both `p1` and `p2` have the same type, and both are not `vararg`.

punzki replied:
The test doesn't have anything to do with `vararg`... It's testing that the intention works on annotation call arguments.

I think historically the reason that in `superClassConstructor.kt`, `p2` was chosen instead of `p1` was _before_ `MixedNamedArgumentsInTheirOwnPosition`, you could only have named arguments at the end of the argument list. After that language feature, either could be used, and this is tested in `mixingPositionedAndNamed.kt`.

Anyway, I don't see a really good reason to change this particular test, even though we could go with `p1` or just go with one argument.

jsjeon replied:
I don't expect you to change the test; I'm just curious why `p2` is preferred over `p1`. I think I found an answer. Once we use `p1` with named argument, `p2` can't be positional argument anymore, i.e., `p2` should be named argument as well. And I think that's why all the following tests kinda prefer to pick the later parameter, or... un-named argument just before the named arguments start? :)

punzki replied:
Just to clarify, as I mentioned in my previous comment, `p1` _can_ be named by itself (e.g., `@A(p1 = 1, 2)`) when `MixedNamedArgumentsInTheirOwnPosition` is enabled (since 1.4). But I think `p2` was used since the tests were written before this feature.

punzki committed:
FIR IDE: Add HLAddNameToArgumentIntention to plugin XML.

darthorimar commented on fir/src/org/jetbrains/kotlin/idea/fir/intentions/HLAddNameToArgumentIntention.kt:
```diff
@@ -0,0 +1,108 @@
+            element is KtContainerNode || super.skipProcessingFurtherElementsAfter(element)
+
+    companion object {
+        val applicator = applicator<KtValueArgument, Input> {
```
What do you think about moving this applicator with correpsonding to `org.jetbrains.kotlin.idea.fir.applicators`. So, it can be used from other inspections/intentions/quickfixes

punzki replied:
I'm going to re-use it in the next set of changes (`AddNamesToCallArgumentsIntention / AddNamesToFollowingArgumentsIntention`) and I can move it then.

punzki committed:
IDE: Move `myFixture.configureByFiles()` call inside
`withCustomCompilerOptions()` in AbstractIntentionTest.

Fixture configuration can cause resolution, which causes language
version settings to be cached for the project and module. This should
only happen after all the custom compiler options are read.

darthorimar commented:
Merged manually
